content/posts/manage-list-feature-flags-with-file/index.md

```diff
@@ -0,0 +1,227 @@
++++
+date = "2021-12-13T14:28:23+07:00"
+author = "nhattien.ho"
+description = "Manage list feature flags with file yaml/json."
+title = "Create feature flags using file in Unleash"
+categories = ["Feature toggle", "Testing"]
+tags = ["Feature toggle", "Unleash", "Node.js"]
+slug = "manage-list-flags-with-file"
++++
+
+In this blog, we will learn how to create a list of feature toggles in Unleash using file. First, let's learn some concept of feature toggle and Unleash 
+#### What is feature toggle (feature flag)?
+Feature toggle (Feature flag) is a technique that allows you to disable some functionality of your application, through settings, without having to deploy new code.
+#### What is Unleash?
+[Unleash](https://docs.getunleash.io/) is a feature toggle system, that gives you a great overview of all feature toggles across all your applications and services. It comes with official client implementations for Java, Node.js, Go, Ruby, Python and .Net.
+#### Why do we use file to create feature flags instead of Admin UI?
+In Manabie, we manage the addition and removal of feature flags with pull request, so we need a solution to manage the list of feature flags by file. Developers can see and change the list of feature flags by modifying this file, done't need to access the admin page. The list of feature flags will be updated after we deploy new code backend.
+#### How to use file to manage list of feature flags in Unleash?
+To run Unleash locally, we need:
+* [Node.js](https://nodejs.org/en/download/) (version 14 or later)
+* [PostgreSQL](https://www.postgresql.org/download/) (version 10 or later)
+* [Create an unleash user and database.](https://docs.getunleash.io/deploy/database-setup)
+##### Start Unleash server
+* Create a new folder/directory on your development computer.
+* From a terminal/bash shell, install the dependencies:
+``` bash
+ npm init
+ npm install unleash-server --save
+```
+* Create a file called server.js, paste the following into it and save.
+```javascript
+const unleash = require('unleash-server');
+let options = {
+    db: {
+        ssl: false,
+        host: 'localhost',
+        port: 9432,
+        database: 'unleash',
+        user: 'root',
+        password: 'root',
+    },
+    server: {
+        port: 4242,
+    },
+}
+unleash
+  .start(options)
+  .then((unleash) => {
+    console.log(
+      `Unleash started on http://localhost:${unleash.app.get('port')}`,
+    );
+  });
+```
+* Run `server.js`:
+```bash
+node server.js
```

Comment:
I'm having some problems running `node server.js`

```
┌─[tuananh@tuananh-manabie:~/Desktop/manabie/unleash]
└─╼ node server.js
/home/tuananh/Desktop/manabie/unleash/server.js:1
const unleash = require(unleash-server);
                        ^

ReferenceError: Cannot access 'unleash' before initialization
    at Object.<anonymous> (/home/tuananh/Desktop/manabie/unleash/server.js:1:25)
    at Module._compile (node:internal/modules/cjs/loader:1101:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1153:10)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12)
    at node:internal/main/run_main_module:17:47
```

```
┌─[tuananh@tuananh-manabie:~/Desktop/manabie/unleash]
└─╼ npm init
This utility will walk you through creating a package.json file.
It only covers the most common items, and tries to guess sensible defaults.

See `npm help init` for definitive documentation on these fields
and exactly what they do.

Use `npm install <pkg>` afterwards to install a package and
save it as a dependency in the package.json file.

Press ^C at any time to quit.
package name: (unleash) 
version: (1.0.0) 
description: 
git repository: 
keywords: 
author: 
license: (ISC) 
About to write to /home/tuananh/Desktop/manabie/unleash/package.json:

{
  "name": "unleash",
  "version": "1.0.0",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1",
    "start": "node server.js"
  },
  "author": "",
  "license": "ISC",
  "dependencies": {
    "unleash-server": "^4.4.4"
  },
  "devDependencies": {},
  "description": ""
}


Is this OK? (yes) 
┌─[tuananh@tuananh-manabie:~/Desktop/manabie/unleash]
└─╼  npm install unleash-server --save

up to date, audited 308 packages in 2s

11 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities
```

Reply:
```sh
┌─[tuananh@tuananh-manabie:~/Desktop/manabie/unleash]
└─╼ node --version
v16.13.1
```

Reply:
Because you write `const unleash = require(unleash-server);`, but `const unleash = require('unleash-server');` is correct, you miss ' ' in `require`

Reply:
rightttt that explains it